cmake/SetupCompilerOptions.cmake

```diff
@@ -86,6 +86,10 @@ elseif("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang") # For Clang or AppleClang
     set(COMPILER_FAMILY_IS_CLANG 1)
     message(STATUS "Compiler family is Clang")
     
+elseif("${CMAKE_CXX_COMPILER_ID}" MATCHES "HCC") # For ROCm
```

Comment:
Which version of CMake is required to correctly detect the compiler ID?

Reply:
I admit to being a bit confused about this variable.  It is not defined by any CMake rev (that I know of).  I force a redefinition in https://github.com/chipfreitag/blt/blob/feature/rocm/cmake/thirdparty/FindROCm.cmake

but I don't think it actually controls anything.  HCC is a clang compiler, but often the default  clang behaviours are not appropriate for building a rocm app.

Reply:
Not defined for HCC that is...

Reply:
@chipfreitag 

CMake will define the CMAKE_CXX_COMPILER_ID -- I suspect in this case it just reports CLANG, and you are overriding to provide logic to select proper flags throughout CMake. You are also overriding the compiler paths, linker paths, etc. 

This is tricky, the goal is to do DEPENDS_ON rocm and have all the magic propgate? 
CMake doesn't know too much about HCC, so I think you *have* to rewire things including the linker, etc?

To help me understand:  It seems HCC will be used to compile everything and there are extra settings for when we have a rocm dep?

If this is the case, we should try to get away with modifying the minimal set of builtin CMake vars. 

I think it makes sense to refactor like @kennyweiss suggested, passing in hcc as the compiler (via any of the cmake mechanisms, or a host-config file). Assuming it survives all of CMake's built-in detection logic, we then have BLT detect we are using hcc as the main compiler and finish the overrides. 

When we are sure its hcc, we can create some variable that indicates that we really have hcc, and we can key off of this in the BLT compiler setup macro stuff, instead of overriding CMAKE_CXX_COMPILER_ID.

You will have to override some things (like the linker command) But flags, includes, and libs,  can be handled by  BLT_REG_LIB for rocm. Again, not sure what we can get away with, but we should aim for the minimal set of overrides. 

I am happy to help with the process, but testing is won't be easy for me. 

Does this make sense, what do folks think?